### DISCUSSION TITLE
套路 - Miao Yu | 于淼

### DISCUSSION THREAD

**Discussion by giscus** [yufree/yufree.cn#146] [General]
# 套路 - Miao Yu | 于淼

https://yufree.cn/cn/2023/05/11/routine/

**Comment by hongtaoh** [comment]
评个论，顺便看一下新的评论系统如何。你的观点我同意。这让我想到高考作文，记得当时最多的素材就是一些人如何坚持然后成功。当然，大家都是为了应考而已。但这些素材以及以个人经历为内容的书，确实太片面了。现在知识爆炸，有太多太多人生道理。这反而成了一个问题，因为在别人身上有用的道理在自己身上还真不一定有用。而且太多名人传记都过大估计了个人努力的作用，而忽视了性格、命运、时代。相比之下，我觉得个人努力才是最微不足道的部分。

**Reply by yufree** [reply]
确实，不过从出版角度，强调这个书就没法卖了，只能去卷个人努力向的奇遇了

**Comment by earfanfan** [comment]
哈哈，以后终于可以来三水这里瞎霍霍啦。

如果2087年的时候，三水本体会消失的话，可以现在考虑一下把这里托付给自己的朋友照看呐。

**Reply by yufree** [reply]
不用麻烦别人了，我到时训练个有我性格的AI，接个太阳能板跟网卡扔到近地轨道上监管就可以，老大哥永远凝视地球。

**Reply by CyrusYip** [reply]
哈哈，我觉得可以存到 https://web.archive.org/ ，然后给它捐钱。

**Reply by earfanfan** [reply]
三水的这个想法抗不过物理上的破坏啊。要是刚好有颗星星把太空上面的东西砸坏了，没有人会去修理。三水可以考虑一下托付给我，因为我的预期寿命更长，等到我狗带了，我再托付益辉，因为益辉有后代，益辉的后代还可能会有后代，哈哈哈，子子孙孙无穷尽，于是这些博主本体已经消亡的博客会一直存在着。

不过，话又说回来，干嘛要存在那么久啊。

**Reply by CyrusYip** [reply]
> 不过，话又说回来，干嘛要存在那么久啊。

大概是想继续活着吧。如果一个人死了，认识他的人也死了，那他就被遗忘了。人都会死，但有作品留下就可以继续「活着」，给后世留下些精神财富。有天我在想我死了之后要怎么办身后事，那会就很想把博客留下来，花钱请人续费域名、维护博客，最好还能印成书，悄悄地藏去各地的图书馆（我好像在某篇博客看到这个做法），哈哈哈。

不过大多数人都没有可以继续或者的作品吧，真的是宇宙里的一粒灰尘呐。清明扫墓的时候多数墓都没有碑，感觉里面的人已经被遗忘了。

> 三水可以考虑一下托付给我，因为我的预期寿命更长，等到我狗带了，我再托付益辉，因为益辉有后代，益辉的后代还可能会有后代，哈哈哈，子子孙孙无穷尽，于是这些博主本体已经消亡的博客会一直存在着。

现在人工智能进步很大，也许还可以用博客作为素材造一个虚拟的三水。

**Reply by yihui** [reply]
> 最好还能印成书，悄悄地藏去各地的图书馆（我好像在某篇博客看到这个做法）

我在 https://yihui.org/cn/2022/06/useless-prosper/ 的第 4 条提过这样的故事。

**Comment by earfanfan** [comment]
另，虽然三水又在说写博客发泄下怨气撒的，但我还是很不厚道地觉得好好笑啊，尽管也不知道笑点是撒。但是我想问问，为撒会对看不惯的事物产生怨气呢？根据我的个人经验，遇到看不惯的事物，我会产生讨厌的心态。

话说俺的同事里有一个非常敷衍塞责的人，给别人挖坑甩锅无数，我非常讨厌他，但是这个人的本性又是那种温和善良的人。后来我想，社会还是很有包容性的啊，这样的人都能好好地生存下来。再后来我又想，“社会”只是人类在创造文字以后造出来的一个概念，并不是社会很有包容性，而是社会没有实体，“社会”这个概念根本就和什么人生存与否没有关联，这样的人能好好地生存下来仅仅就是这样的人并没有轻易放弃自己、也想好好生存下去罢了。于是乎，我很看不惯职场上那些责任心弱的同事，但是尊重他们的存在，把他们当做路边的石头。

又另，说起读书的目的，你总结了好几点，不想浪费时间、不想盲从、不想思想被控制、不想放弃对新知的掌控、担心自己落后等等。哈哈，这么多没有一条说中我的情况。为撒做一件事之前非得十分理智地思考那么多的目的，要先有目的才去做事呢？（ps这里我说的做事的范围仅限于业余时间做的一些事）我读书就是因为对想读的那本书产生了兴趣啊，被激发了好奇心。这跟掌控新的知识、担心自己落后没有任何关系。另外，我总觉得人对自己的评价应该是多维度的，至少不应局限于世俗的什么高低、前后。人先、人后这么单一的评价，也太二维了吧。

这里我要多说明一下，我可半点都没有想要改变三水你的意思啊。只是因为我感知到咱们有差异，所以陈述一下我的想法。也一点没有想要跟你吵架的意思，更加不会明知有差异却还非得叫你认同我的。

又又另，关于“主义”和“意识形态”，我想你并不是反感这两个词。举个例子，一个叫做马克思的人，将他的思想写下来，后人觉得随便称呼马克思的思想这样好像有点不太高大上啊，于是升了一个台阶，将马克思的思想称之为马克思主义，当马克思的思想被称为马克思主义后，就变成了一面拥有象征意义的旗帜。我想你并不是反感这个叫做马克思的人把自己所思所想写下来，应该是反感一个人的思想被抽象成了一面旗帜，后人时常把这面旗帜挥来舞去，旗帜也并不只有这一面，还有很多人挥舞别的旗帜，你真正反感的是人们挥舞旗帜但是脱离实际。

>事实上我反感各种主义或意识形态，从解决问题角度出发需要的是大胆试错而不是纠结这套思路是哪个门派的，是不是违反了当前“理论”。

又又又另，上个月我给你写信吐槽武汉办马拉松造成了部分区域交通拥堵，你叫我别只冲你抱怨，而是要去做实事。额，我后来找到了武汉的城市留言板，留了言，再后来接到了一通回访电话，给我打电话的人听完我陈述的事后说会帮我记下来的，然后就，就是这样了。再后来我想，个人的权益和能力在集体面前实在有限，如果我真得想做出更大的改变，改变一个城市，除非从政。可是，我又想，一个城市里发生的一件小事令我心生不满，我便怀着怨气想着改变它，倘若是一个国家里发生了什么事令我心生不满，那么我又要想着改变它吗？我都没看清楚它的样子，只见到了一个侧面就想改变它，这样好像是不对的啊。再后来，我又想……太长了，已经絮叨太多了，俺先打住。

**Reply by yufree** [reply]
能回电话就有不错了，这些记录总会有被读到的机会，对任何体制只要允许个体发声就已经合格了。而且从你视角提出的问题在别人那边可能还成了优点，你认为好的改变也可能正好有另一个人打电话抱怨，如果你同时接到两个留言一褒一贬如何做？如果一万条5001条褒4999条贬又如何？体制是不会去思考的，对于个例它只能通过程序延迟回复来收集更多信息，设想下如果每个人的意见都触发社会面实质性改变，这又何尝不是一种独裁呢？而且社会公共资源也不太允许太多朝令夕改，你能看到来自体制的答复大概率是解释性的而不是行动性的。你可以认为自己是对的，但在群体价值判断上体制与个体出发点通常差异很大，你总有发声的自由，但想触发体制改变，个人意志很有限，即使是政治家瞎搞也会被体制反制。当然我说的是现代世俗社会，皇国宗教那些封建余孽体制可能有例外，而且就算封建余孽也得搞点神秘主义理论经典构建权威，这本身也会限制个人权力。

**Reply by earfanfan** [reply]
>你可以认为自己是对的……

这里我得吐个槽，最初给你写信的时候就没有想过对错的问题，后来听了你的建议去留言也并没有这样想。我俩差异这么多，我只觉得有差异好正常啊，根本没想过分对错。虽然我老是强调不是来找你吵架的，但并不是因为我的心态是争个对错，是因为我觉得你的心态是这样……我想的最多的是个体与集体的关系。

假如说，我是一个站在集体对立面的个体，要想依照我的个人意愿去改变集体，那么我需要能够调动或者汇集足够多的力量才能做到。顺着这个思路下去，就理所当然遇到你所描述的事情，即意见差异。那么我面对集体的态度，要么就是被迫接受令我不满的事实，要么继续抗争。但是我的态度显然不是这样的，我始终觉得自己是融入集体中的一个个体，而且思想上、物质上都无法完全独立于集体而存在。虽然说，我这样的个体是集体的一部分，但是就算下一秒我的本体从集体中消失了，这个集体也不会受到丝毫影响，所以，我的立场是观察它啊，而且是身在其中地观察它啊，重点的重点是我自知我的观察只是管中窥豹、不见全貌啊。

好了，阐明了我的基本立场和态度后，我终于想起来，其实最初想抱怨的点是，怎么疫情三年后人们的服从性变高了好多啊喂，以及，怀疑他们在制定什么政策方针的时候有一种思维惯性，大概就是依从“先富带动后富”的那种思维，默认有发展必须有牺牲，根本上就没想过要两全，这次办马拉松默认牺牲的就是那些路拥堵后blablabla不赘述了，正是因为默认牺牲了这一小部分，所以一开始就没想着这一小部分也可以在固定的限制条件下进行改善。

**Reply by earfanfan** [reply]
对了，不知是不是读你的博客读久了的缘故，我看你的回复总会自带一种热心肠好人的滤镜。你的许多文字在我脑补的出发点是，眼看着我要踩坑了，劝谏我不要掉坑里去。如果真得是这样的话，你的劝谏是起了作用的。

**Reply by yufree** [reply]
我建议你不要代入任何滤镜与预设看网上的资料，我只是说你有发表意见的自由，但不要指望体制会对这种意见有多大反馈。可以不确切把体制看成市场，你觉得东西卖贵了可以降价，但别人觉得卖便宜了要涨价，最后展示出的价格是买卖双方多轮博弈的结果，这个过程会比较慢。
另外，集体这个概念本就是虚构出来的，你用对个人的评价方式去评价集体在我看来没多大意义，你可以具体去讨论集体使用的法规合理性，但并不存在集体的实体作为靶子，集体不在乎个人意志因为其本身是整体意志的体现，就好比你大脑意志不在乎具体一个细胞的想法一样。
至于说服从性与牺牲，这都是个人层面对集体的理解与解读，集体根本不关心，集体存在的价值更多是对抗另一个集体而不是个人。如果我再无耻一点，可以说当集体在针对个人时，其更多是杀鸡儆猴对另一个集体表态，具体这个人是谁有啥想法并不重要，当然也可能是为了集体内部思想统一去打压或排除某个个体，但大概率集体对个人本身没有天然的恶意与对立，但集体中某个人对某个人是可能有天然对立的，只是借用了集体作为打压借口。越是现代世俗化的集体，就越是会对内开通疏导渠道，也会对个人意志进行程序限制，牺牲效率换取稳定，这也没有对错之分。

**Reply by earfanfan** [reply]
你回复的内容是我本来就认同的。虽然说，我与你交流的目的并不在于达成两个普通人类之间的互相理解，不过我忽然想多此一举解释一下。

>怀疑他们在制定什么政策方针的时候有一种思维惯性……

我在写这里的时候，所想的“他们”是当权者，具体是指主办马拉松的体育局领导们。集体是集权的，至于集权，“集体对个人没有天然的恶意和对立”，但当权力集中在极少数人手里的时候，天然的忽略是很可能有的，感觉我俩对集权好像已经讨论烂了，这里懒得再展开了。ps看多了集这个字以后，我有点不认识它了。要是继续讨论集体、集权，感觉我们可能要走上“白马非马”的歧路了。正好你在写关于2087的小说，回头我读一下你的小说，然后以对未来的幻想为基础再来找你吐槽这些东西吧。

>如果我再无耻一点，可以说……

话说我读到这里的时候偷着乐了一下，此后每每想起还是忍不住偷着乐。我觉得你描述的观点只是观点，并不是你是“无耻”的，哈哈哈，怎么你的道德标准这么高啊，脑子里想到这些就会说自己无耻……我长这么大好像从来没有在任何时刻冒出来过认为自己无耻的念头。

**Comment by CyrusYip** [comment]
你博客流量大应该是 RSS 文件被下载太多了，不是评论的锅。我博客有个月 959 次访问，就已经用了 17G 流量，多数是 RSS 文件。[我用 Cloudflare 缓存 RSS](https://cyrusyip.org/zh-cn/post/2022/11/22/cache-rss-with-cloudflare/)，效果显著，最近 30 天 Vercel 只用了不到 500M 流量。我觉得 Netlify CDN 套 Cloudflare CDN 不太优雅，可能换成 Cloudflare Pages 比较好。我听说有些 RSS 阅读器的爬虫特别勤快，你的 RSS 文件已经 4.7M 了，用那么多流量也不稀奇。

**Reply by yufree** [reply]
谢谢！终于找到罪魁祸首了，我只改了首页rss最多输出10个，但中文站那边竟然是全输出，等我找时间把那边调整一下，4.7M确实夸张了。

**Reply by yufree** [reply]
我这边改好了，目前rss会全文输出最新10篇全文而不是所有文章rss了，xml文件只有几十KB了，流量耗用断崖式下跌，从一个小时上百兆到现在大概每小时五兆。其实Hugo里是可以配置这个[数目](https://github.com/yufree/yufree.cn/blob/135784c02938a5878dba0ad2ba46d9a96ad9081f/config.yaml#LL11C3-L11C3)且我也配置了，但之前为了让rss输出全文魔改了主题里rss输出[模版](https://github.com/yufree/hugo-lithium-theme/blob/a6940e20a2b2640efbcd5277ff4d8e0b23fd8c85/layouts/_default/rss.xml#L9)，改后模版扔掉了所有配置，导致一直输出所有已经发表文章的全文，也就是那个4.7M，我现在补上了那块配置就可以了。我也建议不去配置RSS缓存，因为rss阅读器一般也会比对xml文件保留最新的输出，所以保留大体积的xml文件其实不太环保，之前文章一般rss阅读器都用自己的缓存服务器缓存了。最后，还是感慨下rss阅读器的勤劳，看意思每个小时拉取二十多次，个人博客一天一次就够了，毕竟绝大多数个人博客一月都不一定有几篇更新。

**Reply by yihui** [reply]
我以前也是超流量了才发现 RSS 文件占据了绝大部分流量，所以后来我大幅精简了 RSS 文件的体积，然后 100Gb 的月度限额现在只需要 5Gb 了；RSS 现在也不是我最耗流量的文件，而是中文日志的列表页，这一个页面占了 5G 里的 1G。Vercel 给出了按流量排序的网址列表，所以哪个页面耗费了多少流量一目了然，解决起来也就有的放矢了。

**Reply by yufree** [reply]
额，netlify这功能不但付费，我交了一个月保护费还看不到具体哪个页面消耗流量大，只能看到页面访问量与流量，里面还没有rss那个文件，到最后只能自己下载网页然后干掉了Disqus。要不是 @CyrusYip 提醒，我怎么也想不到是这个文件的锅，当时我还纳闷Disqus那边明明设置了异步加载按说不该流量这么大，现在看起来就是rss的铁锅了。看意思我可能也要跳船到vercel，之前部署过当备份了。
另外我还发现 Hexo 有个友链朋友圈的[插件](https://github.com/Rock-Candy-Tea/hexo-circle-of-friends)，好像是直接 GitHub Action 自动抓博客的rss，然后有个朋友开了这玩意，很固定一天要抓48次，不过这个插件倒是跟之前我那个半成品 dailyr 差不多，Hugo 啥时候也有个类似插件也挺方便的。

**Comment by earfanfan** [comment]
俺再盖一楼试试。话说之前益辉给我安利换 giscus 的时候说，以后不同楼层之间不会相互打扰，但是自从我留言以后，在我之后的每条留言和回复我都收到邮件提醒了。

**Reply by CyrusYip** [reply]
我也是。

**Reply by yihui** [reply]
我好像也是……

**Reply by earfanfan** [reply]
哈哈，这叫好像也是咩，明明就是啊喂……

**Reply by yihui** [reply]
因为这对我来说是第一个样本，我还需要在别的帖子下验证一下，但估计我当时很可能是搞错了，楼下还是会打扰楼上的。

**Reply by yihui** [reply]
验证完毕，果然任何一楼的回帖都会默认通知整栋楼的人。只能按右边的退订按钮免打扰。Github 这有点不人性。

我自己的 Github 账号设置的通知方式只有网页小蓝点通知，没有邮件，所以还算清静。

**Reply by CyrusYip** [reply]
> 果然任何一楼的回帖都会默认通知整栋楼的人。只能按右边的退订按钮免打扰。Github 这有点不人性。

我觉得默认订阅符合 Issues 和 Discussions 的定位，报告 bug 和提问肯定是想继续收到通知。拿它们作评论区属于不按套路出牌。如果能给部分 Issues 和 Discussions 设置默认不订阅也是很好的。

**Comment by yufree** [comment]
我倒不怀疑那时候还有这个网站，但搞不好那时候互联网已经被人机交互网络取代了变成无人问津的古董交流技术了，还是近地卫星靠谱，加个自进化模块，即使地面技术更新也能自己想办法接入，然后发个公告：欢迎访问
yufree.cn，不要怀疑，那个白痴还在。

**Comment by yihui** [comment]
这 Disqus 评论没搬好啊，只是把数据转发了过来，但并没有建立与日志网页的映射，所以旧评论在旧网页下显示不出来。强烈建议批量修改一下，不然一百步都走完九十步了，就差最后几步，可惜了。

首先，获取所有讨论的标题、正文和 ID（需要事先配置有读写 Discussions 权限的 `GITHUB_TOKEN`）：

```r
get_discussions = function(owner, repo) {
  has_next = TRUE
  next_cursor = NULL
  info = NULL
  while(has_next) {
    query = gh::gh_gql(sprintf('query FindRepo {
      repository(owner: "%s", name: "%s") {
        discussions(first: 100%s) {
          pageInfo {
            startCursor
            endCursor
            hasNextPage
            hasPreviousPage
          }
          edges {
            node {
              title
              body
              id
            }
          }
        }
      }
    }', owner, repo, 
    if (is.null(next_cursor)) '' else paste0(', after: "', next_cursor, '"')
    ))
    res = query$data$repository$discussions
    has_next = res$pageInfo$hasNextPage
    next_cursor = res$pageInfo$endCursor
    info = c(info, lapply(res$edges, function(x) unlist(x$node)))
  }
  info = do.call(rbind, info)
  as.data.frame(info)
}

info = get_discussions('yufree', 'yufree.cn')
```

然后你把这个 `info` 数据框的 `title` 修改为网页标题（从 `info$body` 里应该可以提取），然后 `body` 最后加上 SHA1 码（基于 `title`）。

```r
sha1 = function(x) digest::digest(x, 'sha1', serialize = FALSE)
str_json = function(x) jsonlite::toJSON(x, auto_unbox = TRUE)

for (i in seq_len(nrow(info))) {
  info$title[i] = gsub(...)  # 修改标题，这里是唯一需要你自己做的事情
  info$body[i] = sprintf('%s\n\n<!-- sha1: %s -->', info$body[i], sha1(info$title[i]))
}
```

最后批量更新：

```r
for (i in seq_len(nrow(info))) {
  message(i, ': ', info$title[i])
  print(gh::gh_gql(sprintf('mutation {
  updateDiscussion(input: {discussionId: %s, title: %s, body: %s}) {
    discussion {
      id
      title
    }
  }
}', str_json(info$id[i]), str_json(info$title[i]), str_json(info$body[i]))))
}
```

因为你开启了 Giscus 的严格匹配，所以你最好是验证一下上面 `info$title` 确实与你网页的 `<title>` 标签内容相同。这个可以通过读取 `https://yufree.cn/cn/` HTML 源码来验证，并不难。

另外搬 Disqus 评论时你大概是用了 Rob Hyndman 的脚本，他那个脚本有不少小毛病。你搬过来的评论都变成了代码块（需要拖着横向滚动条阅读，这会逼死读者的），原因可能是四个反引号不当缩进。

`````
        ````
`````

如果是我的话，会先测试搬几条，修正这种明显的小问题了再搬。你也可以考虑干脆重新搬一遍，然后把之前搬过来的删掉。

**Reply by hongtaoh** [reply]
益辉可以弄一个付费服务帮别人评论搬家了，感觉你是这个领域大师了

**Reply by yihui** [reply]
我确实已经成精了，不过这样的客户数量可能用一只手都数得过来，目前我知道的就只剩下你和叶寻了，而你们都懂技术，不需要我帮忙。你的搬起来比较容易，唯一需要注意的是 SHA1 码的计算要用网址，而不是标题；另外就是 Issues 转 Discussions 可以通过预先打标签的方式批量将标签转过去（转完后可以删掉标签），这样会快很多，不用一项一项地转。剩下的事情就是整理 `info` 数据框的内容，最后一个循环修改完所有 Discussions。叶寻的网站会麻烦一些，我前几天问他要不要搬，他说暂时不搬。

**Reply by hongtaoh** [reply]
我是想搬的，但没有那么急切，因为来我这里评论的人总共也没多少，但我会尝试弄一下，giscus 感觉更清爽一些。

**Reply by yufree** [reply]
我确实用了 Rob Hyndman 的脚本，跑了半天我看没动静了就没再管，等下班了我去调整一下，十分感谢！（话说我咋只收到了回复的邮件没收到原评论邮件）

**Reply by yufree** [reply]
搞定，又踩了几个坑，过会写个文章记录下。

**Reply by yihui** [reply]
还有几个小问题：

1. yufree.github.com 的旧链接没有替换过来，例如 https://github.com/yufree/yufree.cn/discussions/155 标题的后缀 `| yufree` 应该替换为 `- Miao Yu | 于淼`（替换后 SHA1 码也要重新计算和更新）
2. https://github.com/yufree/yufree.cn/discussions/151 标题仍然是链接
3. 英文日志的标题仍然缺后缀 `- Miao Yu | 于淼`，如 https://github.com/yufree/yufree.cn/discussions/173
4. 评论正文的开头多了个 `header`，这问题是因为三个反引号后面没有换行导致的，即：

````
```header
Originally posted by ...
```
````

而不应该是

````
```header Originally posted by ...```
````

前三个问题会导致日志和评论无法正确建立映射关系。

**Reply by yufree** [reply]
谢谢！

#151 那个是我晚上做测试用的，已经删除。

旧链接跟英文我后面有空打算集中手动处理下，旧链接是历史遗留问题而英文那边是导出的xml里就缺后缀，数量应该不多。

`header` 那个确实没注意到，我其实是把四引号直接改成单引号了，没注意`header` 本身的意思，这个我后面可以重跑下。

**Reply by yihui** [reply]
现在主要问题都修理完了，就剩我上面说的第一个问题：yufree.github.com 的旧链接没有替换过来，如 https://github.com/yufree/yufree.cn/discussions/287 可以正则表达式替换一下。当然，这个问题在日志页面上看不出来，因为顶楼的内容不会显示出来，只在 Github 上能看到失效的旧链接。

**Reply by yufree** [reply]
一共有六篇，应该改完了，这次我在博客下面回复，看看会不会触发叶寻跟袁凡的提醒。

**Reply by earfanfan** [reply]
会。

**Reply by CyrusYip** [reply]
会。

**Reply by earfanfan** [reply]
看到叶寻同学的回复，我本来只是偷着乐了一下，不过当我明白乐的点是益辉大神翻车了，哈哈，可以明着乐了。有一首诗是这么描述这件事的：

人生到处知何似，应是大神要翻车。
路上偶然灰尘大，翻车哪复计东西。

**Reply by hongtaoh** [reply]
我观看了全程表演，这说明我好像没必要搬到 giscus 了。

**Reply by CyrusYip** [reply]
@hongtaoh 和 utterances 钉子户握个手。
